### PR TITLE
Add `Dataset::has_capability`

### PR DESCRIPTION
**Pull request by Atreyagaurav**
- [X] I agree to follow the project's [code of conduct](https://github.com/georust/gdal/blob/master/CODE_OF_CONDUCT.md).
- [X] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

This is for doing things like `Dataset::has_capability(DatasetCaps::ODsCTransactions)` to check if dataset supports Transaction. 

I added the Capabilities from:
- https://gdal.org/en/latest/doxygen/ogr__api_8h.html#a9010219bbc2e32627064ed860048d979
- https://gdal.org/en/latest/doxygen/gdal_8h.html#a8dc635a536f123e1b03004dc74b4da86

I followed the example of `Layer::has_capability` for both of them.

This will help use the `if {} else {}` block for capability based logic like the one mentioned in #582 

Basically it'll make it so that @lnicola 's Suggestion given below can work:
```rust
let stuffs = |d: &Dataset| -> Result<()> { ... };
if /* transactions are supported */ {
    let txn = d.start_transaction()?;
    stuffs(&txn)?;
    txn.commit()?;
} else {
    stuffs(&d)?;
}
```

Currently, without this function, we have to wait for the result of `Dataset::start_transaction` to know if it supports it, in which case we cannot use `if () {} else {}` block due to Rust's borrow rules as the `Dataset::start_transaction` will take `&mut Dataset`.

### CODE REVIEW

**Comment by Atreyagaurav** [comment]
About the Enum variants, since `clippy` warns about the Prefix, should we just remove them? Since these are rust enum, I think we don't need the prefixes like `ODrC`, `ODsC`, `OLC` etc. on the enum itself as the Enum names (`LayerCaps`, `DriverCaps`, `DatasetCaps` etc.) are enough to distinguish them unlike in C.

I just followed what `LayerCaps` was doing for the other two. Changing that would be breaking change, but at least for the new ones we could not add Prefixes.

File: src/driver.rs [hunk 1]
```diff
@@ -301,6 +321,13 @@ impl Driver {
     pub unsafe fn c_driver(&self) -> GDALDriverH {
         self.c_driver
     }
+
+    pub fn has_capability(&self, capability: DriverCaps) -> bool {
+        unsafe {
+            gdal_sys::OGR_Dr_TestCapability(self.c_driver(), capability.into_cstring().as_ptr())
```
[the remark this answers is not in the document]
Comment (Atreyagaurav) [reply src/driver.rs:327]: Maybe I should remove `Driver::has_capability` in that case. And we can add the new one later. Or accept the legacy API for now. But considering there are only two capability to check, I think it is fine to not have this method for now.

What do you think?

**Comment by Atreyagaurav** [comment]
I'm waiting on whether to remove `Driver::has_capability` since it is based on legacy API, or rewrite it to use the `DCAP` metadata instead. Or maybe just adding a metadata checking function (if it doesn't exist already) is sufficient.

**Comment by lnicola** [comment]
I think we can do something like below, what do you say?

```rust
struct DatasetCapability(&'static CStr); // we could also support non-'static ones, but probably not worth it

impl DatasetCapability {
    pub const CREATE_LAYER: DatasetCapability = DatasetCapability(c"CreateLayer");
    pub const DELETE_LAYER: DatasetCapability = DatasetCapability(c"DeleteLayer");
}

struct Dataset;
impl Dataset {
    pub fn test_capability(&self, cap: DatasetCapability) -> bool {
        _ = cap.0;

        // this is the best dataset
        true
    }
}
```

This is a little shorter, while still allowing custom caps if the user ever needs one.

**Comment by lnicola** [comment]
> The DatasetCapability is not deprecated, and there are a fixed list on gdal side so I think it's fine.

My version has much less code, runs slightly faster, and can still be extended if GDAL adds a new capability, so I still think it's an improvement.

> Thank you, can you merge with failed tests? or should I do something?

I disabled the test, you can rebase.

**Comment by lnicola** [comment]
That wasn't a clean rebase or squash but.. whatever, I guess. Anyway, please capitalize `datasource` in the doc-comments, and add a `.` at the end, they're a bit hard to read right now.

Afterwards I think we can merge this and use it in #591?

**Comment by Atreyagaurav** [comment]
Sorry, I was struggling with rebase since the Valgrind was in the middle of two commits I made. It should be clean one, I undid everything and made a new commit on top.

**Review by lnicola — APPROVED** [review]

**Comment by lnicola** [comment]
Yeah, it's clean now, thanks!

**Comment by Atreyagaurav** [comment]
> Afterwards I think we can merge this and use it in https://github.com/georust/gdal/pull/591?

About that, this one is so that we can manually do the "if supports transaction" part by ourselves.

I think the #591 is a bit different because even if starting transaction fails we want to continue writing to database as the transaction is optional and for speedup process only.

File: src/dataset.rs [hunk 2]
```diff
@@ -1,4 +1,9 @@
-use std::{ffi::CString, ffi::NulError, path::Path, ptr};
+use std::{
+    ffi::NulError,
+    ffi::{CStr, CString},
```
Comment (lnicola) [inline src/dataset.rs:3]: Nit: let's merge these:

```suggestion
    ffi::{CStr, CString, NulError},
```

and add a changelog entry.

File: src/dataset.rs [hunk 3]
```diff
@@ -327,6 +354,10 @@ impl Dataset {
         }
         Ok(transformation)
     }
+
+    pub fn has_capability(&self, capability: DatasetCapability) -> bool {
+        unsafe { gdal_sys::GDALDatasetTestCapability(self.c_dataset(), capability.0.as_ptr()) == 1 }
```
Comment (lnicola) [inline src/dataset.rs:359]: Well, I do hope that `TRUE` is 1 and not `0xFFFF_FFFF` :sweat_smile:.

Comment (Atreyagaurav) [reply src/dataset.rs:359]: That's what test cases are for. It's working for now.
